**3.1 Predict the drug response of all four drugs based on the mutational and copy number data**

In [1]:
import pandas as pd # 
import numpy as np
Drug1_Gene_dataset = pd.read_csv("Drug1_Gene_dataset.csv", index_col = 0)
Drug2_Gene_dataset = pd.read_csv("Drug2_Gene_dataset.csv", index_col = 0)
Drug3_Gene_dataset = pd.read_csv("Drug3_Gene_dataset.csv", index_col = 0)
Drug4_Gene_dataset = pd.read_csv("Drug4_Gene_dataset.csv", index_col = 0)

In [2]:
Drug1_Gene_dataset.loc[:,"Drug1_response"] = Drug1_Gene_dataset["Drug1_response"].apply(lambda x: 0 if x=="Sensitivity" else 1)
Drug2_Gene_dataset.loc[:,"Drug2_response"] = Drug2_Gene_dataset["Drug2_response"].apply(lambda x: 0 if x=="Sensitivity" else 1)
Drug3_Gene_dataset.loc[:,"Drug3_response"] = Drug3_Gene_dataset["Drug3_response"].apply(lambda x: 0 if x=="Sensitivity" else 1)
Drug4_Gene_dataset.loc[:,"Drug4_response"] = Drug4_Gene_dataset["Drug4_response"].apply(lambda x: 0 if x=="Sensitivity" else 1)

In [3]:
Drug1_Gene_dataset.reset_index(inplace = True,drop = True)
Drug2_Gene_dataset.reset_index(inplace = True,drop = True)
Drug3_Gene_dataset.reset_index(inplace = True,drop = True)
Drug4_Gene_dataset.reset_index(inplace = True,drop = True)

In [4]:
Drug1_Gene_dataset

ABCB1_mut  ABL2_mut  ACACA_mut  ACVR1B_mut  ACVR2A_mut  ADCY1_mut  \
0            0         0          0           0           0          0   
1            0         0          0           0           0          0   
2            0         0          0           0           0          0   
3            0         0          0           0           0          0   
4            1         0          0           0           0          0   
..         ...       ...        ...         ...         ...        ...   
214          0         0          0           0           0          0   
215          0         0          0           0           0          0   
216          0         0          0           0           0          0   
217          0         0          0           0           0          0   
218          0         0          0           0           0          0   

     AFF4_mut  AHCTF1_mut  AHNAK_mut  AKAP9_mut  ...  loss:cnaPANCAN417  \
0           0           0          0          0  ...                  0   
1           0           0          0          0  ...                  0   
2           0           0          0          0  ...                  0   
3           0           0          0          0  ...                  0   
4           0           1          1          0  ...                  0   
..        ...         ...        ...        ...  ...                ...   
214         0           0          0          0  ...                  0   
215         0           0          1          0  ...                  0   
216         0           0          0          0  ...                  0   
217         0           1          0          0  ...                  0   
218         0           0          0          0  ...                  0   

     loss:cnaPANCAN418  loss:cnaPANCAN419  loss:cnaPANCAN420  \
0                    0                  0                  0   
1                    0                  0                  0   
2                    0                  0                  0   
3                    0                  0                  0   
4                    0                  0                  0   
..                 ...                ...                ...   
214                  0                  0                  0   
215                  0                  0                  0   
216                  0                  0                  0   
217                  0                  0                  0   
218                  0                  0                  0   

     loss:cnaPANCAN421  loss:cnaPANCAN422  loss:cnaPANCAN423  \
0                    0                  0                  0   
1                    0                  0                  0   
2                    0                  0                  0   
3                    0                  0                  0   
4                    0                  0                  0   
..                 ...                ...                ...   
214                  0                  0                  0   
215                  0                  0                  0   
216                  0                  0                  0   
217                  0                  0                  0   
218                  0                  0                  0   

     loss:cnaPANCAN424  loss:cnaPANCAN425  Drug1_response  
0                    0                  0               0  
1                    0                  0               0  
2                    0                  0               1  
3                    0                  0               1  
4                    0                  0               1  
..                 ...                ...             ...  
214                  0                  0               0  
215                  0                  0               1  
216                  0                  0               1  
217                  0                  0               1  
218              

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# Encoding the target variable
label_encoder = LabelEncoder()
Drug1_Gene_dataset['Drug1_response'] = label_encoder.fit_transform(Drug1_Gene_dataset['Drug1_response'])

# Separating features and target variable
X_1 = Drug1_Gene_dataset.drop(['Drug1_response'], axis=1)
y_1 = Drug1_Gene_dataset['Drug1_response']
X_1.shape

(219, 735)

In [6]:
Drug1_Gene_dataset

ABCB1_mut  ABL2_mut  ACACA_mut  ACVR1B_mut  ACVR2A_mut  ADCY1_mut  \
0            0         0          0           0           0          0   
1            0         0          0           0           0          0   
2            0         0          0           0           0          0   
3            0         0          0           0           0          0   
4            1         0          0           0           0          0   
..         ...       ...        ...         ...         ...        ...   
214          0         0          0           0           0          0   
215          0         0          0           0           0          0   
216          0         0          0           0           0          0   
217          0         0          0           0           0          0   
218          0         0          0           0           0          0   

     AFF4_mut  AHCTF1_mut  AHNAK_mut  AKAP9_mut  ...  loss:cnaPANCAN417  \
0           0           0          0          0  ...                  0   
1           0           0          0          0  ...                  0   
2           0           0          0          0  ...                  0   
3           0           0          0          0  ...                  0   
4           0           1          1          0  ...                  0   
..        ...         ...        ...        ...  ...                ...   
214         0           0          0          0  ...                  0   
215         0           0          1          0  ...                  0   
216         0           0          0          0  ...                  0   
217         0           1          0          0  ...                  0   
218         0           0          0          0  ...                  0   

     loss:cnaPANCAN418  loss:cnaPANCAN419  loss:cnaPANCAN420  \
0                    0                  0                  0   
1                    0                  0                  0   
2                    0                  0                  0   
3                    0                  0                  0   
4                    0                  0                  0   
..                 ...                ...                ...   
214                  0                  0                  0   
215                  0                  0                  0   
216                  0                  0                  0   
217                  0                  0                  0   
218                  0                  0                  0   

     loss:cnaPANCAN421  loss:cnaPANCAN422  loss:cnaPANCAN423  \
0                    0                  0                  0   
1                    0                  0                  0   
2                    0                  0                  0   
3                    0                  0                  0   
4                    0                  0                  0   
..                 ...                ...                ...   
214                  0                  0                  0   
215                  0                  0                  0   
216                  0                  0                  0   
217                  0                  0                  0   
218                  0                  0                  0   

     loss:cnaPANCAN424  loss:cnaPANCAN425  Drug1_response  
0                    0                  0               0  
1                    0                  0               0  
2                    0                  0               1  
3                    0                  0               1  
4                    0                  0               1  
..                 ...                ...             ...  
214                  0                  0               0  
215                  0                  0               1  
216                  0                  0               1  
217                  0                  0               1  
218              

In [7]:
pca_1 = PCA(n_components=0.95,random_state = 40)  
X_1_pca = pca_1.fit_transform(X_1)
X_1_pca.shape, pca_1.n_components_

((219, 128), 128)

In [8]:
print(pca_1.explained_variance_ratio_)

[0.11745975 0.05133703 0.03468075 0.03423399 0.02924899 0.02761915
 0.02423359 0.02289718 0.02080106 0.01858332 0.01800346 0.01744276
 0.01677601 0.01511363 0.01398592 0.0133182  0.0129825  0.01286986
 0.0124125  0.01172383 0.01112878 0.01077156 0.0104549  0.0100003
 0.00952336 0.00933519 0.00894261 0.00879091 0.00829687 0.00816917
 0.00807065 0.00781441 0.00764758 0.00760712 0.00733553 0.00696372
 0.00683264 0.00661729 0.00642739 0.00626783 0.00612669 0.00593259
 0.00585571 0.00570873 0.00550898 0.00546643 0.00540932 0.0052565
 0.00506001 0.00498898 0.00494596 0.00476493 0.00466709 0.00460699
 0.00459229 0.00452416 0.00434243 0.00425355 0.00421062 0.00415122
 0.00402101 0.00397545 0.0038836  0.00381973 0.00371053 0.0036621
 0.0035668  0.00353988 0.00347483 0.0034004  0.0033687  0.00325747
 0.00324557 0.00316113 0.00309667 0.00304111 0.0029868  0.00294856
 0.00290098 0.00285511 0.00279768 0.00276768 0.0026203  0.00259061
 0.002577   0.00250438 0.00245481 0.00244236 0.00242269 0.0023728

In [9]:
loadings_1 = pca_1.components_.T
loadings_1

array([[ 0.00323949, -0.00142767, -0.01711814, ..., -0.00339885,
         0.02235248, -0.00583411],
       [ 0.00101009, -0.00326571,  0.00167805, ...,  0.04276579,
        -0.01174652, -0.02828991],
       [ 0.00378417,  0.00504664, -0.00503126, ...,  0.00779573,
         0.01269839, -0.00389361],
       ...,
       [-0.        ,  0.        , -0.        , ..., -0.        ,
        -0.        ,  0.        ],
       [-0.        ,  0.        , -0.        , ..., -0.        ,
        -0.        ,  0.        ],
       [ 0.0032146 , -0.00385924, -0.00023539, ...,  0.02452866,
        -0.00570467,  0.00042524]])

In [10]:
features_1 = X_1.columns
loadings_df_1 = pd.DataFrame(loadings_1, index=features_1)
loadings_df_1

0         1         2         3         4         5    \
ABCB1_mut          0.003239 -0.001428 -0.017118 -0.012518 -0.005400  0.040520   
ABL2_mut           0.001010 -0.003266  0.001678 -0.004160  0.001395 -0.001784   
ACACA_mut          0.003784  0.005047 -0.005031  0.002249 -0.006077 -0.002303   
ACVR1B_mut        -0.002967 -0.000948 -0.008914 -0.006517  0.001434  0.030006   
ACVR2A_mut        -0.001649 -0.002146 -0.009359 -0.000391  0.000995  0.017853   
...                     ...       ...       ...       ...       ...       ...   
loss:cnaPANCAN421 -0.000000  0.000000 -0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN422 -0.000000  0.000000 -0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN423 -0.000000  0.000000 -0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN424 -0.000000  0.000000 -0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN425  0.003215 -0.003859 -0.000235 -0.009989 -0.004771 -0.003033   

                        6         7             8         9    ...       118  \
ABCB1_mut         -0.026088 -0.015401  6.077763e-03  0.001424  ...  0.014153   
ABL2_mut           0.003848  0.000142 -7.432862e-03 -0.012078  ...  0.023282   
ACACA_mut          0.003415  0.004987  1.069583e-02 -0.000188  ...  0.013720   
ACVR1B_mut        -0.020709 -0.024312  5.795872e-08  0.003692  ...  0.006361   
ACVR2A_mut        -0.007315 -0.004249 -6.032898e-04  0.003220  ... -0.009913   
...                     ...       ...           ...       ...  ...       ...   
loss:cnaPANCAN421  0.000000 -0.000000  0.000000e+00  0.000000  ...  0.000000   
loss:cnaPANCAN422  0.000000 -0.000000  0.000000e+00  0.000000  ...  0.000000   
loss:cnaPANCAN423  0.000000 -0.000000  0.000000e+00  0.000000  ...  0.000000   
loss:cnaPANCAN424  0.000000 -0.000000  0.000000e+00  0.000000  ...  0.000000   
loss:cnaPANCAN425  0.013571 -0.003266 -8.305796e-03 -0.005162  ...  0.018430   

                        119       120       121       122       123       124  \
ABCB1_mut         -0.006820  0.006444  0.021916 -0.013809 -0.018610  0.027300   
ABL2_mut          -0.012516  0.035060  0.026628  0.010029  0.032973  0.013678   
ACACA_mut         -0.008398  0.004288  0.023306 -0.010742 -0.023762  0.012612   
ACVR1B_mut        -0.003886 -0.013669  0.010248 -0.003382  0.005882  0.002525   
ACVR2A_mut        -0.005062  0.021446 -0.009421 -0.001478  0.010783 -0.006392   
...                     ...       ...       ...       ...       ...       ...   
loss:cnaPANCAN421  0.000000  0.000000  0.000000 -0.000000  0.000000  0.000000   
loss:cnaPANCAN422  0.000000  0.000000  0.000000 -0.000000  0.000000  0.000000   
loss:cnaPANCAN423  0.000000  0.000000  0.000000 -0.000000  0.000000  0.000000   
loss:cnaPANCAN424  0.000000  0.000000  0.000000 -0.000000  0.000000  0.000000   
loss:cnaPANCAN425 -0.007100  0.036869  0.014228  0.047262  0.014794  0.033511   

                        125       126       127  
ABCB1_mut         -0.003399  0.022352 -0.005834  
ABL2_mut           0.042766 -0.011747 -0.028290  
ACACA_mut          0.007796  0.012698 -0.003894  
ACVR1B_mut         0.001478  0.002461 -0.002238  
ACVR2A_mut        -0.005377  0.009304  0.007505  
...                     ...       ...       ...  
loss:cnaPANCAN421 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN422 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN423 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN424 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN425  0.024529 -0.005705  0.000425  

[735 rows x 128 columns]

In [11]:
loadings_df_1 = pd.DataFrame(loadings_1, columns=[f'PC{i+1}' for i in range(pca_1.n_components_)], index=features_1)
loadings_df_1

PC1       PC2       PC3       PC4       PC5       PC6  \
ABCB1_mut          0.003239 -0.001428 -0.017118 -0.012518 -0.005400  0.040520   
ABL2_mut           0.001010 -0.003266  0.001678 -0.004160  0.001395 -0.001784   
ACACA_mut          0.003784  0.005047 -0.005031  0.002249 -0.006077 -0.002303   
ACVR1B_mut        -0.002967 -0.000948 -0.008914 -0.006517  0.001434  0.030006   
ACVR2A_mut        -0.001649 -0.002146 -0.009359 -0.000391  0.000995  0.017853   
...                     ...       ...       ...       ...       ...       ...   
loss:cnaPANCAN421 -0.000000  0.000000 -0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN422 -0.000000  0.000000 -0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN423 -0.000000  0.000000 -0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN424 -0.000000  0.000000 -0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN425  0.003215 -0.003859 -0.000235 -0.009989 -0.004771 -0.003033   

                        PC7       PC8           PC9      PC10  ...     PC119  \
ABCB1_mut         -0.026088 -0.015401  6.077763e-03  0.001424  ...  0.014153   
ABL2_mut           0.003848  0.000142 -7.432862e-03 -0.012078  ...  0.023282   
ACACA_mut          0.003415  0.004987  1.069583e-02 -0.000188  ...  0.013720   
ACVR1B_mut        -0.020709 -0.024312  5.795872e-08  0.003692  ...  0.006361   
ACVR2A_mut        -0.007315 -0.004249 -6.032898e-04  0.003220  ... -0.009913   
...                     ...       ...           ...       ...  ...       ...   
loss:cnaPANCAN421  0.000000 -0.000000  0.000000e+00  0.000000  ...  0.000000   
loss:cnaPANCAN422  0.000000 -0.000000  0.000000e+00  0.000000  ...  0.000000   
loss:cnaPANCAN423  0.000000 -0.000000  0.000000e+00  0.000000  ...  0.000000   
loss:cnaPANCAN424  0.000000 -0.000000  0.000000e+00  0.000000  ...  0.000000   
loss:cnaPANCAN425  0.013571 -0.003266 -8.305796e-03 -0.005162  ...  0.018430   

                      PC120     PC121     PC122     PC123     PC124     PC125  \
ABCB1_mut         -0.006820  0.006444  0.021916 -0.013809 -0.018610  0.027300   
ABL2_mut          -0.012516  0.035060  0.026628  0.010029  0.032973  0.013678   
ACACA_mut         -0.008398  0.004288  0.023306 -0.010742 -0.023762  0.012612   
ACVR1B_mut        -0.003886 -0.013669  0.010248 -0.003382  0.005882  0.002525   
ACVR2A_mut        -0.005062  0.021446 -0.009421 -0.001478  0.010783 -0.006392   
...                     ...       ...       ...       ...       ...       ...   
loss:cnaPANCAN421  0.000000  0.000000  0.000000 -0.000000  0.000000  0.000000   
loss:cnaPANCAN422  0.000000  0.000000  0.000000 -0.000000  0.000000  0.000000   
loss:cnaPANCAN423  0.000000  0.000000  0.000000 -0.000000  0.000000  0.000000   
loss:cnaPANCAN424  0.000000  0.000000  0.000000 -0.000000  0.000000  0.000000   
loss:cnaPANCAN425 -0.007100  0.036869  0.014228  0.047262  0.014794  0.033511   

                      PC126     PC127     PC128  
ABCB1_mut         -0.003399  0.022352 -0.005834  
ABL2_mut           0.042766 -0.011747 -0.028290  
ACACA_mut          0.007796  0.012698 -0.003894  
ACVR1B_mut         0.001478  0.002461 -0.002238  
ACVR2A_mut        -0.005377  0.009304  0.007505  
...                     ...       ...       ...  
loss:cnaPANCAN421 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN422 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN423 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN424 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN425  0.024529 -0.005705  0.000425  

[735 rows x 128 columns]

**What are the important features?**

In [12]:
important_features_list_1 = []
for i, pc in enumerate(loadings_df_1.columns):
    idx_max = np.argmax(np.abs(loadings_df_1[pc]))
    important_features_list_1.append({
        'PC': pc,
        'Drug1 Important Feature': loadings_df_1.index[idx_max],
        'Loading': loadings_df_1[pc][idx_max]
    })
important_features_df_1 = pd.DataFrame(important_features_list_1)
print(important_features_df_1)

        PC                    Drug1 Important Feature   Loading
0      PC1                          loss:cnaPANCAN338  0.241167
1      PC2                          loss:cnaPANCAN337 -0.172435
2      PC3  gain:cnaPANCAN121 (ACTB,CARD11,PMS2,RAC1)  0.187006
3      PC4                          loss:cnaPANCAN338 -0.153986
4      PC5         loss:cnaPANCAN44 (BMPR1A,FAS,PTEN)  0.157585
..     ...                                        ...       ...
123  PC124                                 ARID1A_mut -0.153336
124  PC125                    loss:cnaPANCAN71 (EXT2)  0.129682
125  PC126                                EWSR1-X_mut -0.169329
126  PC127                                SMARCA4_mut  0.160330
127  PC128                          gain:cnaPANCAN374  0.192285

[128 rows x 3 columns]


In [13]:
# Splitting the dataset into training and testing sets
X_train_1, X_test_1, y_train_1, y_test_1 = train_test_split(X_1_pca, y_1, test_size=0.2, stratify=y_1, random_state=42)
y_train_1.value_counts()

Drug1_response
1    133
0     42
Name: count, dtype: int64

In [14]:
# Applying SMOTE to deal with class imbalance
smote = SMOTE(random_state=42)
X_train_1_balanced, y_train_1_balanced = smote.fit_resample(X_train_1, y_train_1)

In [15]:
import warnings
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.metrics import classification_report
warnings.filterwarnings('ignore')


log_reg = LogisticRegression(class_weight='balanced', random_state=41)  # Initialize the Logistic Regression classifier with class weight balanced
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000], # Regularization strength
    'penalty': ['l1', 'l2', 'elasticnet'],  # Type of regularization, the function is to reduce the overfitting
    'solver': ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'],   # Algorithm for optimization, to find the optimal coefficients based on minimum loss function
    'max_iter': [100, 200, 300, 400, 500],  # Maximum number of iterations taken for the solvers to converge
}
grid_search = GridSearchCV(log_reg, param_grid, cv=10, scoring='accuracy', verbose = 0) # Initialize the grid search with 10 fold cross-validation and accuracy as the scoring metric, verbose means to print the progress

grid_search.fit(X_train_1_balanced, y_train_1_balanced) # Fit the model , Train the classifier

# Perform cross-validation and calculate accuracy
#cv_scores = cross_val_score(log_reg, X_train_1_balanced, y_train_1_balanced, cv=10, scoring='accuracy')

# Predict on the test set
y_pred_1 = grid_search.predict(X_test_1)
best_score = grid_search.best_score_
best_score

0.9096866096866097

In [16]:
from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score, accuracy_score

precision_1 = precision_score(y_test_1, y_pred_1)
recall_1 = recall_score(y_test_1, y_pred_1)
f1_score_1 = f1_score(y_test_1, y_pred_1)
auc_1 = roc_auc_score(y_test_1, y_pred_1)
test_accuracy_1 = accuracy_score(y_test_1, y_pred_1)
print("Drug1 response prediction results:")
print("Precision is", precision_1)
print("recall is", recall_1)
print("f1_score is", f1_score_1)
print("AUC_score is", auc_1)
print("Accuracy is", test_accuracy_1)

Drug1 response prediction results:
Precision is 0.9354838709677419
recall is 0.8787878787878788
f1_score is 0.90625
AUC_score is 0.8484848484848485
Accuracy is 0.8636363636363636


In [17]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# Encoding the target variable
label_encoder = LabelEncoder()
Drug2_Gene_dataset['Drug2_response'] = label_encoder.fit_transform(Drug2_Gene_dataset['Drug2_response'])

# Separating features and target variable
X_2 = Drug2_Gene_dataset.drop(['Drug2_response'], axis=1)
y_2 = Drug2_Gene_dataset['Drug2_response']
X_2.shape

(224, 735)

In [18]:
pca_2 = PCA(n_components=0.95, random_state = 42)  
X_2_pca = pca_2.fit_transform(X_2)
X_2_pca.shape, pca_2.n_components_

((224, 129), 129)

In [19]:
loadings_2 = pca_2.components_.T
loadings_2

array([[ 0.00275955, -0.00415638, -0.01117798, ..., -0.01464919,
         0.00288145, -0.01096495],
       [ 0.00077972, -0.00387236,  0.00437637, ..., -0.00148734,
        -0.04948095,  0.01940674],
       [ 0.00357258,  0.00443249, -0.00398356, ..., -0.00414246,
         0.00219215, -0.012852  ],
       ...,
       [ 0.0062757 ,  0.00457745, -0.00314642, ...,  0.00761428,
        -0.0048716 , -0.0171659 ],
       [ 0.        , -0.        , -0.        , ..., -0.        ,
        -0.        ,  0.        ],
       [ 0.00275441, -0.00528262,  0.0064835 , ..., -0.04844728,
         0.0213938 ,  0.03545997]])

In [20]:
features_2 = X_2.columns
features_2

Index(['ABCB1_mut', 'ABL2_mut', 'ACACA_mut', 'ACVR1B_mut', 'ACVR2A_mut',
       'ADCY1_mut', 'AFF4_mut', 'AHCTF1_mut', 'AHNAK_mut', 'AKAP9_mut',
       ...
       'loss:cnaPANCAN416', 'loss:cnaPANCAN417', 'loss:cnaPANCAN418',
       'loss:cnaPANCAN419', 'loss:cnaPANCAN420', 'loss:cnaPANCAN421',
       'loss:cnaPANCAN422', 'loss:cnaPANCAN423', 'loss:cnaPANCAN424',
       'loss:cnaPANCAN425'],
      dtype='object', length=735)

In [21]:
loadings_df_2 = pd.DataFrame(loadings_2, index=features_2)
loadings_df_2 = pd.DataFrame(loadings_2, columns=[f'PC{i+1}' for i in range(pca_2.n_components_)], index=features_2)

In [22]:
loadings_df_2

PC1       PC2       PC3       PC4       PC5       PC6  \
ABCB1_mut          0.002760 -0.004156 -0.011178 -0.019781  0.016996  0.037320   
ABL2_mut           0.000780 -0.003872  0.004376 -0.003491 -0.000696 -0.003063   
ACACA_mut          0.003573  0.004432 -0.003984 -0.000632 -0.001538  0.004287   
ACVR1B_mut        -0.002834 -0.001626 -0.007821 -0.011363  0.012258  0.022320   
ACVR2A_mut        -0.001589 -0.002358 -0.009684 -0.005762  0.008619  0.011849   
...                     ...       ...       ...       ...       ...       ...   
loss:cnaPANCAN421  0.000000 -0.000000 -0.000000  0.000000 -0.000000  0.000000   
loss:cnaPANCAN422  0.006276  0.004577 -0.003146 -0.005516 -0.026778 -0.008880   
loss:cnaPANCAN423  0.006276  0.004577 -0.003146 -0.005516 -0.026778 -0.008880   
loss:cnaPANCAN424  0.000000 -0.000000 -0.000000  0.000000 -0.000000  0.000000   
loss:cnaPANCAN425  0.002754 -0.005283  0.006484 -0.008069 -0.005250  0.000598   

                        PC7       PC8       PC9      PC10  ...     PC120  \
ABCB1_mut          0.019778  0.004100  0.008329  0.003060  ...  0.008756   
ABL2_mut          -0.002199  0.004871 -0.005050 -0.001846  ... -0.020547   
ACACA_mut         -0.010479  0.003465  0.012773  0.003662  ...  0.006378   
ACVR1B_mut         0.027759  0.005810  0.002378  0.002508  ...  0.002784   
ACVR2A_mut         0.008593  0.002961 -0.002348  0.000086  ...  0.002523   
...                     ...       ...       ...       ...  ...       ...   
loss:cnaPANCAN421 -0.000000  0.000000  0.000000 -0.000000  ... -0.000000   
loss:cnaPANCAN422  0.023844  0.011592 -0.020292  0.004086  ... -0.008482   
loss:cnaPANCAN423  0.023844  0.011592 -0.020292  0.004086  ... -0.008482   
loss:cnaPANCAN424 -0.000000  0.000000  0.000000 -0.000000  ... -0.000000   
loss:cnaPANCAN425 -0.006133  0.013507 -0.000612 -0.005207  ... -0.002166   

                      PC121     PC122     PC123     PC124     PC125     PC126  \
ABCB1_mut         -0.049105  0.012050  0.019324 -0.021087  0.012969  0.028705   
ABL2_mut          -0.007318  0.012974  0.047224  0.026105  0.018923 -0.000045   
ACACA_mut         -0.020901  0.014198  0.012241 -0.018860  0.012522 -0.007369   
ACVR1B_mut        -0.011398  0.006066 -0.008610  0.008756 -0.004427  0.007242   
ACVR2A_mut         0.001017 -0.004024  0.006335  0.005771  0.005126 -0.003957   
...                     ...       ...       ...       ...       ...       ...   
loss:cnaPANCAN421 -0.000000  0.000000  0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN422  0.021174  0.007865  0.004863  0.015396  0.004867  0.008448   
loss:cnaPANCAN423  0.021174  0.007865  0.004863  0.015396  0.004867  0.008448   
loss:cnaPANCAN424 -0.000000  0.000000  0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN425  0.030299 -0.009763 -0.000283 -0.018388 -0.016034 -0.007077   

                      PC127     PC128     PC129  
ABCB1_mut         -0.014649  0.002881 -0.010965  
ABL2_mut          -0.001487 -0.049481  0.019407  
ACACA_mut         -0.004142  0.002192 -0.012852  
ACVR1B_mut        -0.001418 -0.000379  0.006362  
ACVR2A_mut         0.000500 -0.010219  0.003364  
...                     ...       ...       ...  
loss:cnaPANCAN421 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN422  0.007614 -0.004872 -0.017166  
loss:cnaPANCAN423  0.007614 -0.004872 -0.017166  
loss:cnaPANCAN424 -0.000000 -0.000000  0.000000  
loss:cnaPANCAN425 -0.048447  0.021394  0.035460  

[735 rows x 129 columns]

In [23]:
important_features_list_2 = []
for i, pc in enumerate(loadings_df_2.columns):
    idx_max = np.argmax(np.abs(loadings_df_2[pc]))
    important_features_list_2.append({
        'PC': pc,
        'Drug2 Important Feature': loadings_df_2.index[idx_max],
        'Loading': loadings_df_2[pc][idx_max]
    })
important_features_df_2 = pd.DataFrame(important_features_list_2)
print(important_features_df_2)

        PC                    Drug2 Important Feature   Loading
0      PC1                          loss:cnaPANCAN338  0.229642
1      PC2                          loss:cnaPANCAN337 -0.198229
2      PC3                   gain:cnaPANCAN177 (IL7R)  0.179573
3      PC4         loss:cnaPANCAN307 (CHD3,GPS2,TP53)  0.158104
4      PC5         loss:cnaPANCAN44 (BMPR1A,FAS,PTEN)  0.143955
..     ...                                        ...       ...
124  PC125                                  CASP8_mut -0.156835
125  PC126  gain:cnaPANCAN395 (AKT1,HSP90AA1,PPP2R5C)  0.172587
126  PC127                                    NF2_mut -0.176811
127  PC128                                  PSIP1_mut  0.164386
128  PC129                                  CASP8_mut  0.160573

[129 rows x 3 columns]


In [24]:
# Splitting the dataset into training and testing sets
X_train_2, X_test_2, y_train_2, y_test_2 = train_test_split(X_2_pca, y_2, test_size=0.2, stratify=y_2, random_state=42)

# Applying SMOTE to deal with class imbalance
smote = SMOTE(random_state=41)
X_train_2_balanced, y_train_2_balanced = smote.fit_resample(X_train_2, y_train_2)

In [25]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.metrics import classification_report

# Initialize the RandomForest classifier with class weight balanced
random_forest = RandomForestClassifier(class_weight='balanced', random_state=42)
param_grid_2 = {
    'n_estimators': [10, 50, 100, 200],  
    'max_features': ['auto', 'sqrt', 'log2'],  
    'max_depth': [None, 10, 20, 30, 40, 50],  
    'min_samples_split': [2, 5, 10]  
}
grid_search_2 = GridSearchCV(random_forest, param_grid_2, cv=10, scoring='accuracy', verbose = 0)
# Train the classifier
grid_search_2.fit(X_train_2_balanced, y_train_2_balanced)

# Perform cross-validation and calculate accuracy
#cv_scores = cross_val_score(log_reg, X_train_1_balanced, y_train_1_balanced, cv=10, scoring='accuracy')

# Predict on the test set
y_pred_2 = grid_search_2.predict(X_test_2)
best_score = grid_search.best_score_
best_score

0.9096866096866097

In [26]:
from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score, accuracy_score

precision_2 = precision_score(y_test_2, y_pred_2)
recall_2 = recall_score(y_test_2, y_pred_2)
f1_score_2 = f1_score(y_test_2, y_pred_2)
auc_2 = roc_auc_score(y_test_2, y_pred_2)
test_accuracy_2 = accuracy_score(y_test_2, y_pred_2)
print("Drug2 response prediction results:")
print("Precision is", precision_2)
print("recall is", recall_2)
print("f1_score is", f1_score_2)
print("AUC_score is", auc_2)
print("Accuracy is", test_accuracy_2)

Drug2 response prediction results:
Precision is 0.7619047619047619
recall is 0.9696969696969697
f1_score is 0.8533333333333334
AUC_score is 0.5681818181818182
Accuracy is 0.7555555555555555


In [27]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE
from sklearn.decomposition import PCA

# Encoding the target variable
label_encoder = LabelEncoder()
Drug3_Gene_dataset['Drug3_response'] = label_encoder.fit_transform(Drug3_Gene_dataset['Drug3_response'])

# Separating features and target variable
X_3 = Drug3_Gene_dataset.drop(['Drug3_response'], axis=1)
y_3 = Drug3_Gene_dataset['Drug3_response']
X_3.shape

(231, 735)

In [28]:
pca_3 = PCA(n_components=0.95, random_state = 43)  
X_3_pca = pca_3.fit_transform(X_3)
X_3_pca.shape, pca_3.n_components_

((231, 133), 133)

In [29]:
loadings_3 = pca_3.components_.T
features_3 = X_3.columns
loadings_df_3 = pd.DataFrame(loadings_3, index=features_3)
loadings_df_3 = pd.DataFrame(loadings_3, columns=[f'PC{i+1}' for i in range(pca_3.n_components_)], index=features_3)
loadings_df_3

PC1       PC2       PC3       PC4       PC5       PC6  \
ABCB1_mut          0.002282 -0.003404 -0.002960  0.029788  0.016252  0.021498   
ABL2_mut           0.000541 -0.003899 -0.004660 -0.001519 -0.000496 -0.001678   
ACACA_mut          0.003260  0.003967  0.001306  0.004478 -0.002158  0.006148   
ACVR1B_mut        -0.002659 -0.000706  0.000039  0.017778  0.011384  0.009740   
ACVR2A_mut        -0.001603 -0.001530  0.003342  0.014117  0.007819  0.005472   
...                     ...       ...       ...       ...       ...       ...   
loss:cnaPANCAN421  0.000000  0.000000  0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN422  0.005774  0.004065  0.002155  0.000973 -0.027960 -0.009755   
loss:cnaPANCAN423  0.005774  0.004065  0.002155  0.000973 -0.027960 -0.009755   
loss:cnaPANCAN424  0.000000  0.000000  0.000000 -0.000000 -0.000000  0.000000   
loss:cnaPANCAN425  0.002181 -0.005643 -0.010012  0.001681 -0.005576  0.002533   

                        PC7       PC8       PC9      PC10  ...     PC124  \
ABCB1_mut         -0.028817  0.010634  0.006415  0.011683  ... -0.011239   
ABL2_mut           0.002941  0.002282  0.000442 -0.007293  ... -0.002550   
ACACA_mut          0.006511 -0.006832 -0.003608  0.011238  ...  0.010257   
ACVR1B_mut        -0.026536  0.019433  0.007677  0.007515  ...  0.002118   
ACVR2A_mut        -0.008219  0.007133  0.004442 -0.000765  ... -0.001333   
...                     ...       ...       ...       ...  ...       ...   
loss:cnaPANCAN421  0.000000  0.000000  0.000000 -0.000000  ...  0.000000   
loss:cnaPANCAN422 -0.009148  0.021781  0.012630 -0.013387  ... -0.012508   
loss:cnaPANCAN423 -0.009148  0.021781  0.012630 -0.013387  ... -0.012508   
loss:cnaPANCAN424  0.000000  0.000000  0.000000 -0.000000  ...  0.000000   
loss:cnaPANCAN425  0.007674  0.005909 -0.002558 -0.006187  ...  0.033320   

                      PC125     PC126     PC127     PC128     PC129     PC130  \
ABCB1_mut         -0.043870  0.026974 -0.027194 -0.020400  0.016605  0.006324   
ABL2_mut          -0.007541  0.033703  0.035167  0.025941 -0.005182 -0.027649   
ACACA_mut         -0.036469  0.015385 -0.015803 -0.003184  0.018818 -0.001619   
ACVR1B_mut        -0.001101  0.008664  0.001164 -0.003229  0.002908  0.003717   
ACVR2A_mut         0.002363  0.006324  0.011019  0.003568 -0.001344 -0.001141   
...                     ...       ...       ...       ...       ...       ...   
loss:cnaPANCAN421 -0.000000  0.000000  0.000000  0.000000 -0.000000  0.000000   
loss:cnaPANCAN422  0.002333 -0.015059  0.026795  0.027761 -0.000598  0.002601   
loss:cnaPANCAN423  0.002333 -0.015059  0.026795  0.027761 -0.000598  0.002601   
loss:cnaPANCAN424 -0.000000  0.000000  0.000000  0.000000 -0.000000  0.000000   
loss:cnaPANCAN425  0.020096 -0.019228 -0.027107 -0.023802  0.039104 -0.003772   

                      PC131     PC132     PC133  
ABCB1_mut         -0.027979 -0.016871 -0.045098  
ABL2_mut           0.020539  0.013071 -0.050048  
ACACA_mut         -0.006677 -0.020527 -0.041620  
ACVR1B_mut        -0.000556 -0.003048 -0.003668  
ACVR2A_mut        -0.008452 -0.001263  0.004731  
...                     ...       ...       ...  
loss:cnaPANCAN421  0.000000 -0.000000 -0.000000  
loss:cnaPANCAN422 -0.017029 -0.012307 -0.019269  
loss:cnaPANCAN423 -0.017029 -0.012307 -0.019269  
loss:cnaPANCAN424  0.000000 -0.000000 -0.000000  
loss:cnaPANCAN425 -0.049936 -0.041795  0.015871  

[735 rows x 133 columns]

In [30]:
important_features_list_3 = []
for i, pc in enumerate(loadings_df_3.columns):
    idx_max = np.argmax(np.abs(loadings_df_3[pc]))
    important_features_list_3.append({
        'PC': pc,
        'Drug3 Important Feature': loadings_df_3.index[idx_max],
        'Loading': loadings_df_3[pc][idx_max]
    })
important_features_df_3 = pd.DataFrame(important_features_list_3)
print(important_features_df_3)

        PC                    Drug3 Important Feature   Loading
0      PC1                          loss:cnaPANCAN338  0.221366
1      PC2                          loss:cnaPANCAN337 -0.218166
2      PC3                           loss:cnaPANCAN26 -0.173084
3      PC4  gain:cnaPANCAN121 (ACTB,CARD11,PMS2,RAC1) -0.159839
4      PC5                           loss:cnaPANCAN92 -0.163289
..     ...                                        ...       ...
128  PC129                   gain:cnaPANCAN165 (CDK4) -0.158083
129  PC130                   gain:cnaPANCAN344 (MYCN)  0.171976
130  PC131                          loss:cnaPANCAN343 -0.155792
131  PC132                  gain:cnaPANCAN269 (MACF1) -0.170310
132  PC133                  loss:cnaPANCAN30 (LARP4B) -0.145233

[133 rows x 3 columns]


In [31]:
# Splitting the dataset into training and testing sets
X_train_3, X_test_3, y_train_3, y_test_3 = train_test_split(X_3_pca, y_3, test_size=0.2, stratify=y_3, random_state=42)

# Applying SMOTE to deal with class imbalance
smote = SMOTE(random_state=42)
X_train_3_balanced, y_train_3_balanced = smote.fit_resample(X_train_3, y_train_3)

In [32]:
# import warnings
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score, GridSearchCV
# warnings.filterwarnings('ignore')
# Initialize the Logistic Regression classifier with class weight balanced
svc = SVC(class_weight='balanced', random_state=43)
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000],
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
    'gamma': ['scale', 'auto'],
    'max_iter': [100, 200, 300, 400, 500]
}
grid_search_3 = GridSearchCV(svc, param_grid, cv=10, scoring='accuracy')
# Train the classifier
grid_search_3.fit(X_train_3_balanced, y_train_3_balanced)

# Perform cross-validation and calculate accuracy
#cv_scores = cross_val_score(log_reg, X_train_1_balanced, y_train_1_balanced, cv=10, scoring='accuracy')

# Predict on the test set
y_pred_3 = grid_search_3.predict(X_test_3)
best_score = grid_search_3.best_score_
best_score

0.9022486772486772

In [33]:
from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score, accuracy_score

precision_3 = precision_score(y_test_3, y_pred_3)
recall_3 = recall_score(y_test_3, y_pred_3)
f1_score_3 = f1_score(y_test_3, y_pred_3)
auc_3 = roc_auc_score(y_test_3, y_pred_3)
test_accuracy_3 = accuracy_score(y_test_3, y_pred_3)
print("Drug3 response prediction results:")
print("Precision is", precision_3)
print("recall is", recall_3)
print("f1_score is", f1_score_3)
print("AUC_score is", auc_3)
print("Accuracy is", test_accuracy_3)

Drug3 response prediction results:
Precision is 0.7142857142857143
recall is 0.8571428571428571
f1_score is 0.7792207792207793
AUC_score is 0.42857142857142855
Accuracy is 0.6382978723404256


In [34]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# Encoding the target variable
label_encoder = LabelEncoder()
Drug4_Gene_dataset['Drug4_response'] = label_encoder.fit_transform(Drug4_Gene_dataset['Drug4_response'])

# Separating features and target variable
X_4 = Drug4_Gene_dataset.drop(['Drug4_response'], axis=1)
y_4 = Drug4_Gene_dataset['Drug4_response']
X_4.shape

(214, 735)

In [35]:
pca_4 = PCA(n_components=0.95, random_state = 4)  
X_4_pca = pca_4.fit_transform(X_4)
X_4_pca.shape, pca_4.n_components_

((214, 127), 127)

In [36]:
loadings_4 = pca_4.components_.T
features_4 = X_4.columns
loadings_df_4 = pd.DataFrame(loadings_4, index=features_3)
loadings_df_4 = pd.DataFrame(loadings_4, columns=[f'PC{i+1}' for i in range(pca_4.n_components_)], index=features_4)
loadings_df_4

PC1           PC2           PC3           PC4  \
ABCB1_mut          6.486409e-03 -1.332915e-03 -7.292060e-03 -6.624875e-03   
ABL2_mut           1.042877e-03 -3.473925e-03  2.345113e-03 -3.997519e-03   
ACACA_mut          3.886756e-03  4.981430e-03 -6.277027e-03 -5.614246e-04   
ACVR1B_mut         3.469447e-18 -1.110223e-16 -5.551115e-17  1.110223e-16   
ACVR2A_mut        -1.595706e-03 -2.309076e-03 -8.764283e-03  1.208299e-04   
...                         ...           ...           ...           ...   
loss:cnaPANCAN421 -0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
loss:cnaPANCAN422 -0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
loss:cnaPANCAN423 -0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
loss:cnaPANCAN424 -0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
loss:cnaPANCAN425  3.336858e-03 -4.208836e-03  7.357632e-04 -1.221689e-02   

                            PC5           PC6           PC7           PC8  \
ABCB1_mut          9.904442e-03  4.911334e-03  1.093466e-02 -4.449758e-03   
ABL2_mut          -2.401023e-03  1.848397e-03 -5.242795e-03 -2.657673e-03   
ACACA_mut          7.399267e-03 -5.815636e-03 -4.835970e-03 -8.202207e-03   
ACVR1B_mut        -4.466913e-17 -1.665335e-16  6.938894e-17 -1.509209e-16   
ACVR2A_mut         3.345492e-03  1.256807e-02  1.298945e-02  8.929568e-03   
...                         ...           ...           ...           ...   
loss:cnaPANCAN421 -0.000000e+00 -0.000000e+00  0.000000e+00 -0.000000e+00   
loss:cnaPANCAN422 -0.000000e+00 -0.000000e+00  0.000000e+00 -0.000000e+00   
loss:cnaPANCAN423 -0.000000e+00 -0.000000e+00  0.000000e+00 -0.000000e+00   
loss:cnaPANCAN424 -0.000000e+00 -0.000000e+00  0.000000e+00 -0.000000e+00   
loss:cnaPANCAN425  1.435930e-03  2.010994e-03 -1.087197e-02 -3.448507e-03   

                            PC9          PC10  ...         PC118  \
ABCB1_mut         -5.923866e-03  4.152014e-03  ... -4.863712e-02   
ABL2_mut           7.910721e-04  1.328432e-02  ... -2.325410e-02   
ACACA_mut         -9.199284e-04 -3.138818e-04  ... -2.992005e-02   
ACVR1B_mut         2.498002e-16  1.387779e-16  ...  2.428613e-16   
ACVR2A_mut        -3.611146e-03 -1.679124e-03  ...  6.720973e-03   
...                         ...           ...  ...           ...   
loss:cnaPANCAN421  0.000000e+00 -0.000000e+00  ...  0.000000e+00   
loss:cnaPANCAN422  0.000000e+00 -0.000000e+00  ...  0.000000e+00   
loss:cnaPANCAN423  0.000000e+00 -0.000000e+00  ...  0.000000e+00   
loss:cnaPANCAN424  0.000000e+00 -0.000000e+00  ...  0.000000e+00   
loss:cnaPANCAN425  4.302880e-03  7.195034e-03  ... -3.898599e-02   

                          PC119         PC120         PC121         PC122  \
ABCB1_mut         -8.804580e-03  1.565546e-02 -4.573691e-02 -6.682694e-03   
ABL2_mut           4.443632e-02  3.571157e-03 -1.870597e-02 -6.530095e-03   
ACACA_mut         -9.290761e-03  1.678160e-03 -4.751229e-02  8.061792e-03   
ACVR1B_mut        -1.040834e-17 -1.395368e-16  2.480655e-16  1.006140e-16   
ACVR2A_mut        -1.046499e-03  5.138275e-03  1.333380e-02 -8.753799e-03   
...                         ...           ...           ...           ...   
loss:cnaPANCAN421 -0.000000e+00 -0.000000e+00 -0.000000e+00 -0.000000e+00   
loss:cnaPANCAN422 -0.000000e+00 -0.000000e+00 -0.000000e+00 -0.000000e+00   
loss:cnaPANCAN423 -0.000000e+00 -0.000000e+00 -0.000000e+00 -0.000000e+00   
loss:cnaPANCAN424 -0.000000e+00 -0.000000e+00 -0.000000e+00 -0.000000e+00   
loss:cnaPANCAN425  1.037458e-02  4.797884e-02  1.898264e-02 -2.624664e-02   

                          PC123         PC124         PC125         PC126  \
ABCB1_mut          4.727981e-02  1.379219e-02 -2.223313e-02  2.178688e-02   
ABL2_mut          -5.994635e-02 -2.002962e-02  1.649491e-03  1.774259e-02   
ACACA_mut          2.346695e-02 -5.933911e-04 -1.297633e-02 -9.069158e-03   
ACVR1B_mut        -2.706169e-16 -2.081668e-16  1.734723e-18  9.020562e-17   
ACVR2A_mut        -6.316860e-03 -4.915618e-03 -7.75

In [37]:
important_features_list_4 = []
for i, pc in enumerate(loadings_df_4.columns):
    idx_max = np.argmax(np.abs(loadings_df_4[pc]))
    important_features_list_4.append({
        'PC': pc,
        'Drug4 Important Feature': loadings_df_4.index[idx_max],
        'Loading': loadings_df_4[pc][idx_max]
    })
important_features_df_4 = pd.DataFrame(important_features_list_4)
print(important_features_df_4)

        PC                    Drug4 Important Feature   Loading
0      PC1                          loss:cnaPANCAN338  0.239156
1      PC2                          loss:cnaPANCAN337 -0.182218
2      PC3  gain:cnaPANCAN121 (ACTB,CARD11,PMS2,RAC1)  0.199494
3      PC4                           loss:cnaPANCAN26 -0.173905
4      PC5                 loss:cnaPANCAN310 (MAP2K4)  0.171797
..     ...                                        ...       ...
122  PC123                                   CDH1_mut  0.131924
123  PC124                                 ARID1A_mut  0.161611
124  PC125                                   ASPM_mut -0.164340
125  PC126                          loss:cnaPANCAN343  0.151037
126  PC127              gain:cnaPANCAN108 (CHD9,CYLD)  0.151792

[127 rows x 3 columns]


In [38]:
# Splitting the dataset into training and testing sets
X_train_4, X_test_4, y_train_4, y_test_4 = train_test_split(X_4, y_4, test_size=0.2, stratify=y_4, random_state=42)

# Applying SMOTE to deal with class imbalance
smote = SMOTE(random_state=42)
X_train_4_balanced, y_train_4_balanced = smote.fit_resample(X_train_4, y_train_4)

In [39]:
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score, GridSearchCV

# Initialize the SVC classifier with class weight balanced
svc = SVC(class_weight='balanced', random_state=44)
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000],
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
    'gamma': ['scale', 'auto'],
    'max_iter': [100, 200, 300, 400, 500]
}
grid_search_4 = GridSearchCV(svc, param_grid, cv=10, scoring='accuracy')
# Train the classifier
grid_search_4.fit(X_train_4_balanced, y_train_4_balanced)

# Perform cross-validation and calculate accuracy
#cv_scores = cross_val_score(log_reg, X_train_1_balanced, y_train_1_balanced, cv=10, scoring='accuracy')

# Predict on the test set
y_pred_4 = grid_search_4.predict(X_test_4)
best_score = grid_search_4.best_score_
best_score

0.7981666666666667

In [40]:
from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score, accuracy_score

precision_4 = precision_score(y_test_4, y_pred_4)
recall_4 = recall_score(y_test_4, y_pred_4)
f1_score_4 = f1_score(y_test_4, y_pred_4)
auc_4 = roc_auc_score(y_test_4, y_pred_4)
test_accuracy_4 = accuracy_score(y_test_4, y_pred_4)
print("Drug4 response prediction results:")
print("Precision is", precision_4)
print("recall is", recall_4)
print("f1_score is", f1_score_4)
print("AUC_score is", auc_4)
print("Accuracy is", test_accuracy_4)

Drug4 response prediction results:
Precision is 0.7272727272727273
recall is 0.7741935483870968
f1_score is 0.75
AUC_score is 0.5120967741935484
Accuracy is 0.627906976744186


**3.2 Which drugs can you predict better than random?**

 I can predict drug1, drug2，drug3 and drug4 better than random.

**3.3 What are the important features?**

The important features of different Drugs are above.

 **3.4 How could you improve your predictions?**
 
1. Implement the Synthetic Minority Over-sampling Technique (SMOTE) to address sample imbalance issues, thereby augmenting the minority class through synthetic data generation.

2. Employ Stratified K-fold cross-validation to ensure that each fold is a representative of the overall distribution, thus mitigating distribution imbalances across the data splits.

3. Utilize Grid Search for hyperparameter optimization, systematically exploring a range of configurations to find the most effective model parameters.

4. Apply Principal Component Analysis (PCA) to identify the principal features, reducing dimensionality while retaining the variance in the dataset.

5. Increase the number of folds in K-fold cross-validation to enhance the robustness of the model evaluation, ensuring a more reliable estimation of model performance.

6. Explore alternative machine learning algorithms such as Support Vector Machines (SVM), Random Forest, and others to determine the most suitable model for the given dataset.

7. Implement ensemble methods like Boosting and Stacking to leverage the collective strength of multiple models, potentially improving prediction accuracy through their synergistic capabilities.